File: app/src/main/java/com/thewizrd/simpleweather/radar/EarthWindMapViewProvider.java
package com.thewizrd.simpleweather.radar;

import android.os.Build;
import android.os.Bundle;
import android.view.ViewGroup;
import android.webkit.RenderProcessGoneDetail;
import android.webkit.WebSettings;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.fragment.app.Fragment;

import com.thewizrd.shared_resources.utils.AnalyticsLogger;
import com.thewizrd.shared_resources.utils.Colors;
import com.thewizrd.shared_resources.utils.StringUtils;
import com.thewizrd.shared_resources.utils.WeatherUtils;
import com.thewizrd.simpleweather.helpers.RadarWebClient;
import com.thewizrd.simpleweather.helpers.WebViewHelper;

import java.util.Locale;

@RequiresApi(value = Build.VERSION_CODES.LOLLIPOP)
public class EarthWindMapViewProvider extends RadarViewProvider {
    private String radarURL;

    public EarthWindMapViewProvider(@NonNull Fragment fragment, @NonNull ViewGroup rootView) {
        super(fragment, rootView);
    }

    @Override
    public void updateCoordinates(@NonNull WeatherUtils.Coordinate coordinates, boolean updateView) {
        radarURL = String.format(Locale.ROOT, RadarProvider.EARTHWINDMAP_URL_FORMAT, coordinates.getLongitude(), coordinates.getLatitude());
        if (updateView) updateRadarView();
    }

    @Override
    public void updateRadarView() {
        WebView webView = getRadarWebView();

        if (webView == null) {
            getViewContainer().addView(webView = createWebView());
        }

        if (interactionsEnabled()) {
            WebViewHelper.enableInteractions(webView);
        } else {
            WebViewHelper.disableInteractions(webView);
        }

        if (!StringUtils.isNullOrWhitespace(radarURL)) {
            WebViewHelper.loadUrl(webView, radarURL);
        } else {
            WebViewHelper.loadUrl(webView, RadarProvider.EARTHWINDMAP_DEFAULT_URL);
        }
    }

    @Override
    public void onDestroyView() {
        if (getViewContainer() != null) {
            WebView webView = getRadarWebView();
            if (webView != null) {
                WebViewHelper.loadBlank(webView);
                getViewContainer().removeAllViews();
                webView.destroy();
            }
        }
    }

    @Override
    public void onResume() {
        if (getViewContainer() != null) {
            WebView webView = getRadarWebView();
            if (webView != null) {
                webView.onResume();
            }
        }
    }

    @Override
    public void onPause() {
        if (getViewContainer() != null) {
            WebView webView = getRadarWebView();
            if (webView != null) {
                webView.onPause();
            }
        }
    }

    @Override
    public void onConfigurationChanged() {
        if (getViewContainer() != null) {
            WebView webView = getRadarWebView();
            if (webView != null) {
                if (radarURL != null) {
                    WebViewHelper.forceReload(webView, radarURL);
                } else {
                    WebViewHelper.forceReload(webView, RadarProvider.EARTHWINDMAP_DEFAULT_URL);
                }
            }
        }
    }

    private WebView getRadarWebView() {
        if (getViewContainer() != null) {
            return (WebView) getViewContainer().getChildAt(0);
        }

        return null;
    }

    @NonNull
    private WebView createWebView() {
        WebView webView = new WebView(getParentFragment().getContext());

        // WebView
        WebViewHelper.restrictWebView(webView);
        WebViewHelper.enableJS(webView, true);
        webView.getSettings().setRenderPriority(WebSettings.RenderPriority.HIGH);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            webView.setRendererPriorityPolicy(WebView.RENDERER_PRIORITY_IMPORTANT, true);
        }

        webView.setWebViewClient(new RadarWebClient(false) {
            @Override
            public boolean onRenderProcessGone(WebView view, RenderProcessGoneDetail detail) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    Bundle args = new Bundle();
                    args.putBoolean("didCrash", detail.didCrash());
                    args.putInt("renderPriorityAtExit", detail.rendererPriorityAtExit());
                    AnalyticsLogger.logEvent("WeatherRadarFragment: render gone", args);
                } else {
                    AnalyticsLogger.logEvent("WeatherRadarFragment: render gone");
                }

                if (getViewContainer() != null) {
                    WebView wv = getRadarWebView();

                    if (wv == view) {
                        getViewContainer().removeAllViews();
                        wv = null;
                        view.loadUrl("about:blank");
                        view.onPause();
                        view.destroy();
                        updateRadarView();
                        return true;
                    }
                }

                return super.onRenderProcessGone(view, detail);
            }
        });
        webView.setBackgroundColor(Colors.BLACK);
        webView.resumeTimers();

        return webView;
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/radar/nullschool/EarthWindMapViewProvider.kt
package com.thewizrd.simpleweather.radar.nullschool

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.ViewGroup
import android.webkit.RenderProcessGoneDetail
import android.webkit.WebSettings
import android.webkit.WebView
import androidx.annotation.RequiresApi
import com.google.android.material.snackbar.Snackbar
import com.thewizrd.shared_resources.utils.AnalyticsLogger
import com.thewizrd.shared_resources.utils.Colors
import com.thewizrd.shared_resources.utils.Coordinate
import com.thewizrd.shared_resources.utils.StringUtils
import com.thewizrd.simpleweather.helpers.RadarWebClient
import com.thewizrd.simpleweather.helpers.WebViewHelper.disableInteractions
import com.thewizrd.simpleweather.helpers.WebViewHelper.enableInteractions
import com.thewizrd.simpleweather.helpers.WebViewHelper.enableJS
import com.thewizrd.simpleweather.helpers.WebViewHelper.forceReload
import com.thewizrd.simpleweather.helpers.WebViewHelper.loadBlank
import com.thewizrd.simpleweather.helpers.WebViewHelper.postLoadUrl
import com.thewizrd.simpleweather.helpers.WebViewHelper.restrictWebView
import com.thewizrd.simpleweather.radar.RadarViewProvider
import java.util.*

@RequiresApi(value = Build.VERSION_CODES.LOLLIPOP)
class EarthWindMapViewProvider(context: Context, rootView: ViewGroup) : RadarViewProvider(context, rootView) {
    companion object {
        private const val EARTHWINDMAP_DEFAULT_URL = "https://earth.nullschool.net/#current/wind/surface/level/overlay=precip_3hr"
        private const val EARTHWINDMAP_URL_FORMAT = "$EARTHWINDMAP_DEFAULT_URL/orthographic=%s,%s,3000"
    }

    private var radarURL: String? = null
    override fun updateCoordinates(coordinates: Coordinate, updateView: Boolean) {
        radarURL = String.format(Locale.ROOT, EARTHWINDMAP_URL_FORMAT, coordinates.longitude, coordinates.latitude)
        if (updateView) updateRadarView()
    }

    override fun onCreateView(savedInstanceState: Bundle?) {
        super.onCreateView(savedInstanceState)
        createWebView()?.let {
            viewContainer.addView(it)
        }
    }

    override fun updateRadarView() {
        var webView = radarWebView
        if (webView == null) {
            createWebView()?.let {
                viewContainer.addView(it)
                webView = it
            }
        }

        if (interactionsEnabled()) {
            webView?.enableInteractions()
        } else {
            webView?.disableInteractions()
        }

        if (!StringUtils.isNullOrWhitespace(radarURL)) {
            webView?.postLoadUrl(radarURL!!)
        } else {
            webView?.postLoadUrl(EARTHWINDMAP_DEFAULT_URL)
        }
    }

    override fun onDestroyView() {
        if (viewContainer != null) {
            radarWebView?.let {
                it.loadBlank()
                viewContainer.removeAllViews()
                it.destroy()
            }
        }
    }

    override fun onResume() {
        if (viewContainer != null) {
            radarWebView?.onResume()
        }
    }

    override fun onPause() {
        if (viewContainer != null) {
            radarWebView?.onPause()
        }
    }

    override fun onConfigurationChanged() {
        if (viewContainer != null) {
            val webView = radarWebView
            if (webView != null) {
                if (radarURL != null) {
                    webView.forceReload(radarURL)
                } else {
                    webView.forceReload(EARTHWINDMAP_DEFAULT_URL)
                }
            }
        }
    }

    private val radarWebView: WebView?
        get() {
            return viewContainer?.getChildAt(0) as WebView?
        }

    private fun createWebView(): WebView? {
        if (!context.packageManager.hasSystemFeature(PackageManager.FEATURE_WEBVIEW)) {
            AnalyticsLogger.logEvent("EarthWindMap: no webview installed")
            Snackbar.make(viewContainer, "WebView not installed", Snackbar.LENGTH_SHORT).show()
            return null
        }

        return WebView(context).apply {
            // WebView
            restrictWebView()
            enableJS(true)

            settings.setRenderPriority(WebSettings.RenderPriority.HIGH)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                setRendererPriorityPolicy(WebView.RENDERER_PRIORITY_IMPORTANT, true)
            }

            webViewClient = object : RadarWebClient(!interactionsEnabled()) {
                override fun onRenderProcessGone(view: WebView, detail: RenderProcessGoneDetail): Boolean {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        val args = Bundle().apply {
                            putBoolean("didCrash", detail.didCrash())
                            putInt("renderPriorityAtExit", detail.rendererPriorityAtExit())
                        }
                        AnalyticsLogger.logEvent("WeatherRadarFragment: render gone", args)
                    } else {
                        AnalyticsLogger.logEvent("WeatherRadarFragment: render gone")
                    }

                    if (viewContainer != null) {
                        var wv = radarWebView

                        if (wv === view) {
                            viewContainer.removeAllViews()
                            wv = null
                            view.loadUrl("about:blank")
                            view.onPause()
                            view.destroy()
                            updateRadarView()
                            return true
                        }
                    }

                    return super.onRenderProcessGone(view, detail)
                }
            }

            setBackgroundColor(Colors.BLACK)
            resumeTimers()
        }
    }
}